### ISSUE TITLE
[$250] mWeb - Onboarding - Onboarding can be dismissed with device back button and website freezes

### ISSUE THREAD

**Issue by vincdargento** [Expensify/App#54473]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 9.0.77-6
**Reproducible in staging?:** Yes
**Reproducible in production?:** Yes
**If this was caught on HybridApp, is this reproducible on New Expensify Standalone?:** N/A
**If this was caught during regression testing, add the test name, ID and link from TestRail:** N/A
**Email or phone of affected tester (no customers):** ibellicotest+115@gmail.com
**Issue reported by:** Applause Internal Team
**Device used:** Motorola MotoG60 - Android 12 - Chrome
**App Component:** Other

## Action Performed:
1. Open the staging.new.expensify.com website.
2. Sign in with a new account.
3. When the onboarding modal is displayed, use device back button twice or more times if necessary.

## Expected Result:
Onboarding modal shouldn´t be dismissed when using device back button.

## Actual Result:
Onboarding modal is dismissed after using device back button and the website freezes on inbox after this action.

## Workaround:
Unknown

## Platforms:
- [ ] Android: Standalone
- [ ] Android: HybridApp
- [x] Android: mWeb Chrome
- [ ] iOS: Standalone
- [ ] iOS: HybridApp
- [ ] iOS: mWeb Safari
- [ ] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

https://github.com/user-attachments/assets/ecf17ce5-4217-428b-9b7c-7bee0bc0f559

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021872156796368987725</li>
        <li>Upwork Job ID: 1872156796368987725</li>
        <li>Last Price Increase: 2025-01-02</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @rushatgabhane</details>

**Comment by Shahidullah-Muffakir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The issue occurs when a user is on the onboarding screen, and the onboarding modal is dismissed after pressing the device back button. 
### What is the root cause of that problem?
We did not handle the back button in the onboarding flow, the device back button press triggers the dismissal of the onboarding modal, which should not happen during the onboarding process.
### What changes do you think we should make in order to solve the problem?
To resolve this issue, we should intercept the back button press while the onboarding modal is displayed. This can be done by:
- Using the `BackHandler` API (for Android) to prevent the default behavior when the back button is pressed.
- Ensuring that the back button press doesn't close the modal, but rather allows the user to stay in the onboarding flow.
in the https://github.com/Expensify/App/blob/8b7096fc4c4d522407a6f51b4fe0fd3543e5feb2/src/pages/OnboardingPurpose/BaseOnboardingPurpose.tsx#L50 we can add this useEffect: 
```javascript
useEffect(() => {
    // Add event listener for back button press
    BackHandler.addEventListener('hardwareBackPress', () => true);
    return () => {
        BackHandler.removeEventListener('hardwareBackPress', () => true);
    };
}, [isFocused]);
```

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?
NA

**Comment by Shahidullah-Muffakir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The issue occurs when a user is on the onboarding screen, and the onboarding modal is dismissed after pressing the device back button.

### What is the root cause of that problem?
We did not handle the back button in the onboarding flow in the `web`, and as a result, the device back button press triggers the dismissal of the onboarding modal. 
The code is added only for native:
https://github.com/Expensify/App/blob/8b7096fc4c4d522407a6f51b4fe0fd3543e5feb2/src/pages/OnboardingPurpose/index.native.tsx#L8

### What changes do you think we should make in order to solve the problem?
To resolve this issue, we should intercept the back button press while the onboarding modal is displayed. This can be done by using the same code as use in the native for web as well
in the https://github.com/Expensify/App/blob/8b7096fc4c4d522407a6f51b4fe0fd3543e5feb2/src/pages/OnboardingPurpose/index.tsx#L6
add the same code:



```javascript
    // To block android native back button behavior
    useFocusEffect(
        useCallback(() => {
            // Return true to indicate that the back button press is handled here
            const backAction = () => true;

            const backHandler = BackHandler.addEventListener('hardwareBackPress', backAction);

            return () => backHandler.remove();
        }, []),
    );
### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?
NA

**Comment by samranahm** [comment]
@Shahidullah-Muffakir 

Event listeners like `BackHandler.addEventListener` work only in the native environment and will not have any effect on the web.

**Comment by techievivek** [comment]
Removing the blocker label since it's related to frontend only.

**Comment by railway17** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
The issue occurs when a user is on the onboarding screen, and the onboarding modal is dismissed after pressing the device back button.

### What is the root cause of that problem?
We did not handle the back button pressing event for the web,

### What changes do you think we should make in order to solve the problem?
- We have to add a new handler to prevent the pressing back button.
- It can be implemented by creating a new custom hook which will be created inside `src/hooks/usePreventBackNavigation.ts`.
- We can use this hook in `BaseOnboardingPurpose` component.

### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?
N/A

**Comment by Senninseyi** [comment]
**Proposal**
<hr>

**Please re-state the problem that we are trying to solve in this issue.**

The onboarding modal on Expensify’s mWeb interface is dismissed when the back button is pressed, causing the website to freeze on the inbox page.

**What is the root cause of that problem?**

- The back button event is not correctly handled, allowing unintended modal dismissal.

- State management fails to maintain stability after the modal is dismissed.

**What changes do you think we should make in order to solve the problem?**

1. Add a listener to prevent the default back button action when the onboarding modal is active.

2. Ensure the application state remains consistent if a back button action occurs during onboarding.

3. Fix transitions and cleanup to avoid app freezes when dismissals occur improperly.

**What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?**

NA

**Comment by Senninseyi** [comment]
Contributor details
Your Expensify account email: seyi.oyebamiji@gmail.com
Upwork Profile Link: https://www.upwork.com/freelancers/~01bdedaf1b63d8c485?mp_source=share

**Comment by ikevin127** [comment]
`BackHandler` only works on native as the [documentation page](https://reactnative.dev/docs/backhandler) states:

> [!warning]
> **Warning for modal users:**
> If your app shows an opened Modal, BackHandler will not publish any events ([see Modal docs](https://reactnative.dev/docs/modal#onrequestclose)).

given that on narrow layout devices we display stack navigator screens like `OnboardingPurpose` as modal (see [OnboardingModalNavigator](https://github.com/Expensify/App/blob/e94df0f79443f99b33be867fec8637c6a9aa5821/src/libs/Navigation/AppNavigator/Navigators/OnboardingModalNavigator.tsx#L71-L74)).

To further reinforce this point, here's a commit [`c87e7d1: fix BackHandler is not supported on web and should not be used. error`](https://github.com/Expensify/App/pull/24482/commits/c87e7d15a17d9629568b3f4534e377c161651394) where we applied a patch in `react-native-modal` to fix an error which would state that `BackHandler` is not supported on web.

##

I spent some time trying to consistently reproduce the issue and was not able to on **Android: mWeb Chrome**, since most of the times if I press the hardware back button, it usually either goes back to empty page (no URL) or it closes the browser app altogether, then when I reopen it and navigate to the app, only then, in some cases I will be stuck on home page which is frozen with URL being `/onboarding/purpose`.

https://github.com/user-attachments/assets/feb6a642-e14c-4c03-a4a2-59a26ad8d8dc

I think we need more datailed steps in order to consistently reproduce the issue as without that we cannot properly debug and propose a fix, it'd only be guesswork.

**Comment by MitchExpensify** [comment]
I don't think this is worth fixing right now. A user who is trying exit by going back multiple times is not a motivated lead
